king_phisher/client/windows/campaign_import.py

```diff
@@ -0,0 +1,439 @@
+import logging
+import threading
+import xml.etree.ElementTree as ET
+
+from king_phisher import utilities
+from king_phisher import serializers
+from king_phisher.client import gui_utilities
+from king_phisher.client.widget import extras
+from king_phisher.errors import KingPhisherInputValidationError
+
+import advancedhttpserver
+from gi.repository import Gtk
+from gi.repository import GLib
+
+__all__ = ('ImportCampaignWindow',)
+
+class ImportThread(threading.Thread):
+	def __init__(self, *args, **kwargs):
+		super(ImportThread, self).__init__(*args, **kwargs)
+		self.stop_flag = threading.Event()
+		self.stop_flag.clear()
+
+	def stop(self):
+		self.stop_flag.set()
+
+	def stopped(self):
+		return self.stop_flag.isSet()
+
+class ImportCampaignWindow(gui_utilities.GladeGObject):
+	"""
+	Display a dialog which allows a new campaign to be created or an
+	existing campaign to be opened.
+	"""
+	dependencies = gui_utilities.GladeDependencies(
+		children=(
+			'progressbar',
+			'textview',
+			'entry_campaign',
+			'entry_file',
+			'button_import',
+			'button_select',
+			'spinner'
+		),
+		top_level=('ImportCampaignWindow',)
+	)
+	top_gobject = 'window'
+
+	def __init__(self, *args, **kwargs):
+		super(ImportCampaignWindow, self).__init__(*args, **kwargs)
+		self.logger = logging.getLogger('KingPhisher.Client.ImportCampaignWindow')
+		self.rpc = self.application.rpc
+		self.button_import_campaign = self.gobjects['button_import']
+		self.button_import_campaign.set_sensitive(False)
+		self.select_button = self.gobjects['button_select']
+		self.entry_campaign_name = self.gobjects['entry_campaign']
+		self.entry_path = self.gobjects['entry_file']
+		self.import_progress = self.gobjects['progressbar']
+		self.spinner = self.gobjects['spinner']
+		self.text_buffer = self.gobjects['textview'].get_buffer()
+
+		# place holders for once an xml file is loaded
+		self.campaign_info = None
+		self.db_campaigns = None
+		self.thread_import_campaign = None
+
+	def _set_text_view(self, string_to_set):
+		self.text_buffer.set_text(string_to_set + "\n")
+
+	def _update_text_view(self, string_to_add):
+		end_iter = self.text_buffer.get_end_iter()
+		self.text_buffer.insert(end_iter, string_to_add + "\n")
+
+	def _update_id(self, element, id_fields, old_id, new_id):
+		"""
+		Iterates through the element and replaces the specified old ID to the new ID in the requested ID fields.
+
+		:param element: Element to iterate over where the old id values can be found.
+		:type element: :py:class:`xml.etree.ElementTree.Element`
+		:param list id_fields: The list of fields to look for old_id.
+		:param old_id: The old id value that has been changed
+		:param new_id: The new id value to set.
+		"""
+		for nods in element.iter():
+			if nods.tag in id_fields and nods.text == old_id:
+				nods.text = new_id
+				# if new_id is none set type to null
+				if new_id is None:
+					nods.attrib['type'] = 'null'
+
+	def _get_keys_values(self, element):
+		keys = []
```

Comment:
Since you were have performance issues with this code, you might get it to run faster if you swapped all three of these lists, keys, rows and values for [collections.deque](https://docs.python.org/2/library/collections.html#collections.deque) instances. Should get faster appends.